provider/internal/index.go

```diff
@@ -132,88 +131,112 @@ func (i *IndexState) Annotate(a infer.Annotator) {
-	name string,
-	input IndexArgs,
-	preview bool,
-) (string, IndexState, error) {
-	state, err := i.Update(ctx, name, IndexState{}, input, preview)
-	return name, state, err
+	req infer.CreateRequest[IndexArgs],
+) (infer.CreateResponse[IndexState], error) {
+	resp, err := i.Update(ctx,
+		infer.UpdateRequest[IndexArgs, IndexState]{
+			ID:      req.Name,
+			Olds:    IndexState{},
+			News:    req.Inputs,
+			Preview: req.Preview,
+		},
+	)
+	return infer.CreateResponse[IndexState]{ID: req.Name, Output: resp.Output}, err
 }
 
 // Update performs `buildx imagetools create` to create a new OCI index /
 // manifest list.
 func (i *Index) Update(
 	ctx context.Context,
-	name string,
-	state IndexState,
-	input IndexArgs,
-	preview bool,
-) (IndexState, error) {
+	req infer.UpdateRequest[IndexArgs, IndexState],
+) (infer.UpdateResponse[IndexState], error) {
+	state, input := req.Olds, req.News
+
 	state.IndexArgs = input
 	state.Ref = input.Tag
 
 	cli, err := i.client(ctx, state, input)
 	if err != nil {
-		return state, err
+		return infer.UpdateResponse[IndexState]{Output: state}, err
 	}
 
-	if preview {
-		return state, nil
+	if req.Preview {
+		return infer.UpdateResponse[IndexState]{Output: state}, nil
 	}
 
-	provider.GetLogger(ctx).Debugf("creating index with tag %s and sources %s", input.Tag, input.Sources)
+	provider.GetLogger(ctx).
+		Debugf("creating index with tag %s and sources %s", input.Tag, input.Sources)
 
 	err = cli.ManifestCreate(ctx, input.isPushed(), input.Tag, input.Sources...)
 	if err != nil {
-		return state, fmt.Errorf("creating: %w", err)
+		return infer.UpdateResponse[IndexState]{Output: state}, fmt.Errorf("creating: %w", err)
 	}
 
-	_, _, state, err = i.Read(ctx, name, input, state)
+	// Read remote manifest information, if it exists.
+	live, err := i.Read(ctx,
+		infer.ReadRequest[IndexArgs, IndexState]{ID: req.ID, Inputs: input, State: state},
+	)
 	if err != nil {
-		return state, fmt.Errorf("reading: %w", err)
+		return infer.UpdateResponse[IndexState]{Output: state}, fmt.Errorf("reading: %w", err)
 	}
-	return state, nil
+	return infer.UpdateResponse[IndexState]{Output: live.State}, nil
 }
 
 func (i *Index) Read(
 	ctx context.Context,
-	name string,
-	input IndexArgs,
-	state IndexState,
-) (string, IndexArgs, IndexState, error) {
+	req infer.ReadRequest[IndexArgs, IndexState],
+) (infer.ReadResponse[IndexArgs, IndexState], error) {
+	state, input := req.State, req.Inputs
+
 	state.IndexArgs = input
 	state.Ref = input.Tag
 
 	if !input.isPushed() {
 		provider.GetLogger(ctx).Debug("skipping read because index was not pushed")
-		return name, input, state, nil // Nothing to read.
+		return infer.ReadResponse[IndexArgs, IndexState]{
+			ID:     req.ID,
+			Inputs: input,
+			State:  state,
+		}, nil // Nothing to read.
 	}
 
 	cli, err := i.client(ctx, state, input)
 	if err != nil {
-		return name, input, state, err
+		return infer.ReadResponse[IndexArgs, IndexState]{
+			ID:     req.ID,
```

Comment:
Just a note that the request/response structs are obviously great for extensibility, but I felt a little off the guard rails when wiring up things like the ID. Feels easy to mess up. There's definitely room for sugar but it's not blocking.